Associate DrawableHitObject with HitObjectLifetimeEntry rather than HitObject

ekrctb opened the pull request:
Several specific logic is required to handle the difference between "pooled" and "non-pooled" DHOs.
Currently, those two code paths are tangled, and sometimes it is not clear what condition and pooled-ness are needed in some logic, leading to confusion.
My goal is to make the difference between two cases factored out, and using the same code for the common logic in a more easy-to-reason way.

As the first step, this PR introduce a new invariant: a DHO is applied to a hit object then `lifetimeEntry` is not null.
In other words, the canonical model of a DHO is a *HitObjectLifetimeEntry*, not a *HitObject*, though this semantics is currently not in the public interface.
`HitObject` and `Result` are merely properties of the `HitObjectLifetimeEntry` with the DHO. The lifetime entry is only one source of the truth.

In this PR's implementation, when a hit object is applied to a DHO without a lifetime entry supplied, a synthetic lifetime entry is created, and the DHO is associated with it.
In my plan, the next step will be using the synthetic entry in `HitObjectContainer` to unify the lifetime management logic, but that would contain a bit of breaking changes and thus I separated this PR.

ekrctb committed:
Make DHO application logic clearer with Entry/HitObject separation

ekrctb committed:
Create synthetic LifetimeEntry for a DHO when not supplied

Now, a DHO is always associated with a HitObjectLifetimeEntry while used.
Result is always stored in the entry, and not in the DHO.

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -184,8 +184,11 @@ protected override void LoadAsyncComplete()
-                Apply(HitObject, lifetimeEntry);
+            if (initialHitObject != null)
+            {
+                Apply(initialHitObject, null);
+                initialHitObject = null;
```
Don't really like this `initialHitObject` song-and-dance. Wouldn't it work better to just create the unmanaged lifetime entry in the ctor and just apply it if it's not null here?

ekrctb replied:
I had to introduce a state of `lifetimeEntry != null && !hasEntryApplied` so it is not "DHO is applied iff `lifetimeEntry != null`" anymore but I think it is fine.

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -204,27 +207,41 @@ protected override void LoadComplete()
-                // Transfer lifetime from the entry.
-                LifetimeStart = lifetimeEntry.LifetimeStart;
-                LifetimeEnd = lifetimeEntry.LifetimeEnd;
+                if (lifetimeEntry.HitObject != hitObject)
+                    throw new InvalidOperationException($"{nameof(HitObjectLifetimeEntry)} has different {nameof(HitObject)} from the specified one.");
```
Not super sure about this intermediate state of things where runtime exceptions are getting thrown on bad input... Sure, obsoleting this method right now in this PR would hurt, but it'll hurt and be a huge mess either way. Would probably rather see a new overload that only accepts the entry in here too? Or, at least it can be added, without the obsoletion of the old one for now...

ekrctb replied:
I would like to wait to deprecate things until a more complete picture of the situation is drawn. I don't want to make breaking changes for just later realizing a different change was better.

ekrctb replied:
That being said, I added public `Apply(HitObjectLifetimeEntry)` and `Apply(HitObject)` as I agreed the state of having no access to the "correct" method is bad.

bdach replied:
Honestly, that non-committal attitude towards deprecating the existing overload is not giving me confidence in approving this change. I personally need to see a cold hard diff with the deprecated variant removed, and as a result see how the existing usages will behave with the new code paths going forward. Otherwise, unfortunately I am nowhere near smart enough to be able to tell just from my head alone if the new overloads are a good direction or not. Maybe the others are, I don't know.

ekrctb replied:
The change required for the deprecation of this overload is fairly trivial, and I just pushed that change.
My first comment above was probably misleading, but it was a comment about my general objective of not making unnecessary breaking changes in PRs, not specific to this change.

bdach commented on osu.Game/Rulesets/Objects/UnmanagedHitObjectEntry.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public readonly DrawableHitObject DrawableHitObject;
+
+        public UnmanagedHitObjectEntry(HitObject hitObject, DrawableHitObject drawableHitObject)
+            : base(hitObject)
+        {
+            DrawableHitObject = drawableHitObject;
+            LifetimeStart = DrawableHitObject.LifetimeStart;
+            LifetimeEnd = DrawableHitObject.LifetimeEnd;
+        }
```
This should really only accept the DHO and read the hitobject through it, no? Accepting the hitobject alongside the DHO like this is error-prone in this context.

Class also needs xmldoc as to what "unmanaged" means for future travelers.

ekrctb replied:
`DHO.HitObject` cannot be obtained here because the entry is not set.

ekrctb committed:
Add doc comment

ekrctb committed:
Now, DHO.lifetimeEntry can be non-null even it is not fully applied

ekrctb committed:
Make single-argument overloead of DHO.Apply public

ekrctb committed:
Ensure a non-null hitobject entry has a non-null Result

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -164,11 +166,15 @@ public abstract class DrawableHitObject : SkinReloadableDrawable
+            if (initialHitObject != null)
+            {
+                lifetimeEntry = new UnmanagedHitObjectEntry(initialHitObject);
+                ensureEntryHasResult();
```
Is this necessary here? Won't eventual application of the entry ensure the existence of the result anyways?

ekrctb replied:
It is not currently necessary.
I thought having to worry about null Result is another source of a bug.
Ideally `HitObjectLifetimeEntry.Result` should be readonly and set in the constructor, but the current code doesn't allow this, and it is a compromise.

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -198,37 +204,45 @@ protected override void LoadComplete()
-                // Copy any existing result from the entry (required for rewind / judgement revert).
-                Result = lifetimeEntry.Result;
-            }
-            else
-                LifetimeStart = HitObject.StartTime - InitialLifetimeOffset;
```
This behaviour seems to have been lost somewhere along the way (if no entry is given, lifetime start is automatically calculated in this manner; I see this being removed but not added back anywhere else) - I believe we have consumers relying on it, so it needs to remain.

ekrctb replied:
I thought it was fine because `LifetimeStart` is still set in `HitObjectLifetimeEntry` but actually it uses `HitObjectLifetimeEntry`'s `InitialLifetimeOffset` which is different from `DrawableHitObject`'s `InitialLifetimeOffset` so it has to be fixed to not accidentally change the behavior.

ekrctb committed:
Rename UnmanagedHitObjectEntry -> SyntheticHitObjectEntry

"Unmanaged" was confusing because its lifetime is still managed by the HitObjectContainer.

ekrctb committed:
Fix wrong InitialLifetimeOffset is used for a non-pooled DHO.

HitObjectLifetimeEntry's InitialLifetimeOffset is different from DrawableHitObject's InitialLifetimeOffset.

ekrctb committed:
Revert "Fix wrong InitialLifetimeOffset is used for a non-pooled DHO."

This reverts commit 67fcfd9d

ekrctb committed:
Fix wrong InitialLifetimeOffset is used for a non-pooled DHO.

HitObjectLifetimeEntry's InitialLifetimeOffset is different from DrawableHitObject's InitialLifetimeOffset.

ekrctb committed:
Deprecate old overload of Apply

peppy commented:
I don't have any issues with this change in isolation, but like @bdach have a bit of a struggle seeing if this is an improvement at the end of the pipe (ie. with any follow-up changes). So I wouldn't be against merging this if, for instance, @smoogipoo sees this as a good route, but otherwise will be waiting to see follow-up PRs in the chain to make a call.

smoogipoo commented:
Without having looked at the code in depth yet, this was actually the path I was hoping to go down from the very beginning.

smoogipoo reviewed (approved):
lgtm!
